packages/expo/cli/start/doctor/ngrok/ExternalModule.ts

```diff
@@ -0,0 +1,194 @@
+  /**
+   * Prefer to install the package globally, this can be overwritten if the function
+   * detects that a locally installed package simply needs an upgrade, default `false`
+   */
+  prefersGlobalInstall?: boolean;
+}
+
+/** Resolves a local or globally installed package, prompts to install if missing. */
+export class ExternalModule<IModule> {
+  private instance: IModule | null = null;
+
+  constructor(
+    /** Project root for checking if the package is installed locally. */
+    private projectRoot: string,
+    /** Info on the external package. */
+    private pkg: {
+      /** NPM package name. */
+      name: string;
+      /** Required semver range, ex: `^1.0.0`. */
+      versionRange: string;
+    },
+    /** A function used to create the installation prompt message. */
+    private promptMessage: (pkgName: string) => string
+  ) {}
+
+  /** Resolve the globally or locally installed instance, or prompt to install. */
+  async resolveAsync({
+    prefersGlobalInstall,
+    ...options
+  }: ResolvePromptOptions = {}): Promise<IModule> {
+    try {
+      return (
+        this.getVersioned() ??
+        this.installAsync({
+          ...options,
+          shouldGloballyInstall: prefersGlobalInstall,
+        })
+      );
+    } catch (error: any) {
+      if (error instanceof ExternalModuleVersionError) {
+        // If the module version in not compliant with the version range,
+        // we should prompt the user to install the package where it already exists.
+        return this.installAsync({
+          ...options,
+          shouldGloballyInstall: error.shouldGloballyInstall ?? prefersGlobalInstall,
+        });
+      }
+      throw error;
+    }
+  }
+
+  /** Prompt the user to install the package and try again. */
+  async installAsync({
+    shouldPrompt = true,
+    autoInstall,
+    shouldGloballyInstall,
+  }: InstallPromptOptions = {}): Promise<IModule> {
+    const packageName = [this.pkg.name, this.pkg.versionRange].join('@');
+    if (!autoInstall) {
+      // Delay the prompt so it doesn't conflict with other dev tool logs
+      await delayAsync(100);
+    }
+    const answer =
+      autoInstall ||
+      (shouldPrompt &&
+        (await confirmAsync({
+          message: this.promptMessage(packageName),
+          initial: true,
+        })));
+    if (answer) {
+      Log.log(`Installing ${packageName}...`);
+
+      // Always use npm for global installs
+      const packageManager = shouldGloballyInstall
+        ? new PackageManager.NpmPackageManager({
+            cwd: this.projectRoot,
+            log: Log.log,
+            silent: !EXPO_DEBUG,
+          })
+        : PackageManager.createForProject(this.projectRoot, {
+            silent: !EXPO_DEBUG,
+          });
+
+      try {
+        if (shouldGloballyInstall) {
+          await packageManager.addGlobalAsync(packageName);
+        } else {
+          await packageManager.addDevAsync(packageName);
```

Comment:
Is it intentional to install any package as a _dev_ dependency here? For ngrok, that makes sense, but this class seems a bit more generic than just that.

Reply:
The purpose of this module is to resolve packages for use in the CLI, this would mean all packages would need to be dev dependencies.